bridges/modules/xcm-bridge-hub-router/src/lib.rs

```diff
@@ -67,148 +100,276 @@ pub const HARD_MESSAGE_SIZE_LIMIT: u32 = 32 * 1024;
-		/// Local XCM channel manager.
-		type LocalXcmChannelManager: XcmChannelStatusProvider;
+
+		/// Resolves a specific `BridgeId` for `dest`, used for identifying the bridge in cases of
+		/// congestion and dynamic fees. If it resolves to `None`, it means no congestion or
+		/// dynamic fees are handled for `dest`.
+		#[pallet::no_default]
+		type BridgeIdResolver: ResolveBridgeId;
+
+		/// Origin of the sibling bridge hub that is allowed to report bridge status.
+		#[pallet::no_default]
+		type BridgeHubOrigin: EnsureOriginWithArg<Self::RuntimeOrigin, BridgeIdOf<Self, I>>;
 
 		/// Additional fee that is paid for every byte of the outbound message.
+		/// See `calculate_message_size_fees` for more details.
 		type ByteFee: Get<u128>;
-		/// Asset that is used to paid bridge fee.
-		type FeeAsset: Get<AssetId>;
+		/// Asset used to pay the `ByteFee`.
+		/// If not specified, the `ByteFee` is ignored.
+		/// See `calculate_fees` for more details.
+		type FeeAsset: Get<Option<AssetId>>;
 	}
 
+	/// An alias for the `BridgeId` of configured `T::BridgeIdResolver`.
+	pub type BridgeIdOf<T, I> = <<T as Config<I>>::BridgeIdResolver as ResolveBridgeId>::BridgeId;
+
 	#[pallet::pallet]
 	pub struct Pallet<T, I = ()>(PhantomData<(T, I)>);
 
 	#[pallet::hooks]
 	impl<T: Config<I>, I: 'static> Hooks<BlockNumberFor<T>> for Pallet<T, I> {
 		fn on_initialize(_n: BlockNumberFor<T>) -> Weight {
-			// if XCM channel is still congested, we don't change anything
-			if T::LocalXcmChannelManager::is_congested(&T::SiblingBridgeHubLocation::get()) {
-				return T::WeightInfo::on_initialize_when_congested()
+			let mut weight_used = Weight::zero();
+
+			// Iterate all uncongested bridges
+			let mut bridges_to_update = Vec::new();
+			let mut bridges_to_remove = Vec::new();
+			for (bridge_id, mut bridge_state) in Bridges::<T, I>::iter() {
+				weight_used.saturating_accrue(T::DbWeight::get().reads(1));
+
+				// if not congested anymore, we can start to decreasing fee factor
+				if !bridge_state.is_congested {
+					let previous_factor = bridge_state.delivery_fee_factor;
+					let new_factor = previous_factor / EXPONENTIAL_FEE_BASE;
+					if new_factor >= MINIMAL_DELIVERY_FEE_FACTOR {
+						bridge_state.delivery_fee_factor = new_factor;
+						bridges_to_update.push((bridge_id, previous_factor, bridge_state));
+					} else {
+						bridges_to_remove.push((bridge_id, previous_factor));
+					}
+				}
 			}
 
-			// if we can't decrease the delivery fee factor anymore, we don't change anything
-			let mut delivery_fee_factor = Self::delivery_fee_factor();
-			if delivery_fee_factor == MINIMAL_DELIVERY_FEE_FACTOR {
-				return T::WeightInfo::on_initialize_when_congested()
+			// remove
+			for (bridge_id, previous_value) in bridges_to_remove.into_iter() {
+				log::info!(
+					target: LOG_TARGET,
+					"Bridge channel with id {:?} is uncongested. Removing fee factor!",
+					bridge_id,
+				);
+				Bridges::<T, I>::remove(&bridge_id);
```

Comment:
Why do we remove a bridge if the fee is too small? is this storage item only tracking bridges with fees?

Reply:
Yes, basically, there’s no need to store the bridge when the factor is going to be lower than `MINIMAL_DELIVERY_FEE_FACTOR`. Also, a factor lower than `MINIMAL_DELIVERY_FEE_FACTOR` (1) could decrease the original fees more than intended.